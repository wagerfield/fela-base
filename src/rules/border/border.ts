import { arrayReduce } from "fast-loops"
import {
  isArray,
  parseEdges,
  wrapKey
} from "../../core/utils"
import { ALL } from "../../core/edges"

export interface BorderStyle {
  [key: string]: any
}

export interface BorderRuleOptions {
  edges?: string | string[]
  width?: string | number
  style?: string
  color?: string
}

export default ({
  edges = ALL,
  color = "black",
  style = "solid",
  width = 1
}: BorderRuleOptions = {}) => {
  const keys = parseEdges(edges)
  if (isArray(keys)) {
    if (keys.length === 4) {
      return {
        borderWidth: width,
        borderStyle: style,
        borderColor: color
      }
    } else {
      const borderStyle: BorderStyle = {}
      return arrayReduce(
        keys,
        (result, key) => {
          result[wrapKey(key, "border", "width")] = width
          result[wrapKey(key, "border", "style")] = style
          result[wrapKey(key, "border", "color")] = color
          return result
        },
        borderStyle
      )
    }
  } else {
    throw new Error("invalid edges value")
  }
}
