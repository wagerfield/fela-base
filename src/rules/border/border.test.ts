import border from "./border"

describe("border", () => {
  it("throws error with invalid edges", () => {
    expect(() => {
      border({ edges: null })
    }).toThrowErrorMatchingSnapshot()
  })

  it("renders expected snapshot with no options", () => {
    expect(border()).toMatchSnapshot()
  })

  it("renders expected snapshot when edges is a string", () => {
    expect(
      border({ edges: "left|right" })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot when edges is an array", () => {
    expect(
      border({ edges: ["top", "bottom"] })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot when styles are set", () => {
    expect(
      border({
        style: "dashed",
        color: "red",
        width: 2
      })
    ).toMatchSnapshot()
  })
})
