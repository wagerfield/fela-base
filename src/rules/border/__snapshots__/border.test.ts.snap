// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`border renders expected snapshot when edges is a string 1`] = `
Object {
  "borderLeftColor": "black",
  "borderLeftStyle": "solid",
  "borderLeftWidth": 1,
  "borderRightColor": "black",
  "borderRightStyle": "solid",
  "borderRightWidth": 1,
}
`;

exports[`border renders expected snapshot when edges is an array 1`] = `
Object {
  "borderBottomColor": "black",
  "borderBottomStyle": "solid",
  "borderBottomWidth": 1,
  "borderTopColor": "black",
  "borderTopStyle": "solid",
  "borderTopWidth": 1,
}
`;

exports[`border renders expected snapshot when styles are set 1`] = `
Object {
  "borderColor": "red",
  "borderStyle": "dashed",
  "borderWidth": 2,
}
`;

exports[`border renders expected snapshot with no options 1`] = `
Object {
  "borderColor": "black",
  "borderStyle": "solid",
  "borderWidth": 1,
}
`;

exports[`border throws error with invalid edges 1`] = `"invalid edges value"`;
